README.md
[![npm version](https://badge.fury.io/js/lp5562.svg)](https://badge.fury.io/js/lp5562)

# LP5562

This package is a quick and dirty assembler for the [Texas Instruments LP5562][lp5562]. Package is primarily written so that an open-source toolchain can be provided for the LED controller used in the [keyswitch.one][kso] project.

## Usage

First install via npm:

    npm install --save lp5562

Then include it in your project

```typescript
// if using typescript or ES6 modules
import {assemble} from 'lp5562'

// or if using ES5
var assemble = require('lp5562').assemble;
```

Typescript definitions are included in the package. The function's signature is:

### `assemble(program : string) : Uint16Array`

where `program` is the string representation of the assembly program, and the resulting `Uint16Array` is an array of 48 16-bit WORDs representing the memory map of LP5562's program space.

```typescript
assemble(`
  .engine1
  set_pwm 255

  flash:
    ramp 200, -128
    ramp 200, 128
    branch 5, flash

  ramp 200, -128
  ramp 200, -127
  end, i
`);
```

Produces:

    Uint16Array[ 65344, 65283, 32515, 33186, 65283, 65027, 208, 0, 0, 0, ... 0 ]

All data is provided in big-endian format (as the LP5562 is an 8-bit device). Feed the `Uint16Array#buffer` to a `Uint8Array` or use a `DataView` if you need a different data representation.

## Grammar

For the assembly language spec, see the [grammar](src/lib/grammar.ne). As a rule of thumb, all instructions that work on the official assembler provided with the evaluation module should work here as well. See §3.4.1 and §3.4.2 in the [EVM docs](docs/snvu203a.pdf) for syntax and parameters.

## Contributing

This project is MIT Licensed. Pull requests welcome. Changes documented in [CHANGELOG.md](CHANGELOG.md).

[lp5562]: http://www.ti.com/product/LP5562
[kso]: http://keyswitch.one


CHANGELOG.md
# Change Log

## [0.1.1] - 2017-06-12

- Re-published to NPM due to user-account shuffling.
- Readme significantly updated


## 0.1.0 - 2017-06-12

- Initial project release
- Single exported `assemble` function for most common use-case for now.

[0.1.1]: https://github.com/sizigi/lp5562/compare/v0.1.0...v0.1.1

src/test/test.ts
import * as assert from 'assert';
import * as fs from 'fs';
import {assemble} from '../lib/index';
import {suite, test} from 'mocha-typescript';
import {readHex, assertEqualBuffer} from './util';

@suite class UtilTest {
  @test bufferSimpleEqual() {
    let a = new Uint8Array(4).fill(0xAA);
    let b = new Uint16Array(2).fill(0xAAAA);
    assertEqualBuffer(a.buffer, b.buffer);
  }

  @test bufferComplexEqual() {
    let a = new Uint8Array([0xDE, 0xAD, 0xBE, 0xEF]);
    let b = new Uint16Array(2);
    let dv = new DataView(b.buffer);
    dv.setUint16(0, 0xDEAD);
    dv.setUint16(2, 0xBEEF);
    assertEqualBuffer(a.buffer, b.buffer);
  }

  @test bufferNotEqual() {
    let a = new Uint8Array([0xDE, 0xAD, 0xBE, 0xEF]);
    let b = new Uint16Array([0xCAFE, 0xBABE]);
    assert.throws(() => assertEqualBuffer(a.buffer, b.buffer));
  }

  @test fileCompare() {
    let a = new Uint8Array(readHex('./src/test/programs/test.hex'));
    let b = new Uint16Array(2);
    let dv = new DataView(b.buffer);

    dv.setUint16(0, 0xBAAD);
    dv.setUint16(2, 0xF00D);

    assertEqualBuffer(a.buffer, b.buffer);
  }
}


suite('ProgramTests', () => {
  for(let i of ['blink', 'breathing', 'color_cycle', 'flash_with_afterglow', 'blink_and_color_change']) {
    test(i, () => {
      let a = readHex(`./src/test/programs/${i}.hex`);
      let b = assemble(fs.readFileSync(`./src/test/programs/${i}.src`).toString());
      assertEqualBuffer(a.buffer, b.buffer);
    });
  }
});



let pgmCases = [
  {name: 'double_label', error: /duplicate label/i},
  {name: 'double_global_label', error: /duplicate label/i},
  {name: 'invalid_program', error: /invalid program/i},
]

suite('BrokenPrograms', () => {
  for(let c of pgmCases) {
    test(c.name, () => {
      let data = fs.readFileSync(`./src/test/cases/${c.name}.src`).toString();
      assert.throws(() => {
        assemble(data);
      }, c.error)
    });
  }
});

  // @test run() {
  //   let a = readHex('./src/test/programs/breathing.hex');
  //   let b = assemble(fs.readFileSync('./src/test/programs/breathing.src').toString());

  //   assertEqualBuffer(a.buffer, b.buffer);


src/test/util.ts
import * as fs from 'fs';
import * as assert from 'assert';
import {buf2hex} from '../lib/util';

export function readHex(fileName: string) : Uint8Array {
  const hex = (fs.readFileSync(fileName)
    .toString()
    .match(/\S+/g) || [])

  const bytes = hex.map(d => parseInt(d, 16));
  return new Uint8Array(bytes);
}

export function assertEqualBuffer(buf1: ArrayBuffer, buf2: ArrayBuffer) {
  assert.equal(0,
    bufferDiffCount(buf1, buf2),
    `Buffers not equal (${buf2hex(buf1)}) and (${buf2hex(buf2)})`
  )
}

export function bufferDiffCount(buf1 : ArrayBuffer, buf2 : ArrayBuffer) : number {
    if (buf1.byteLength != buf2.byteLength) throw new Error(`Buffer length mismatch (${buf1.byteLength} vs ${buf2.byteLength})`);
    var dv1 = new DataView(buf1);
    var dv2 = new DataView(buf2);

    let diff = 0;
    for (var i = 0 ; i < dv1.byteLength; i++) {
        if (dv1.getUint8(i) != dv2.getUint8(i)) diff++;
    }
    return diff;
}


src/lib/grammar.js
// Generated automatically by nearley
// http://github.com/Hardmath123/nearley
(function () {
function id(x) {return x[0]; }

  const empty = d => null;
  const nth = i => (d => d[i]);
  const notnull = d => d != null;
  const v = i => (d => i)
var grammar = {
    Lexer: undefined,
    ParserRules: [
    {"name": "program$ebnf$1", "symbols": []},
    {"name": "program$ebnf$1$subexpression$1", "symbols": ["nl", "line"], "postprocess": nth(1)},
    {"name": "program$ebnf$1", "symbols": ["program$ebnf$1", "program$ebnf$1$subexpression$1"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "program", "symbols": ["line", "program$ebnf$1"], "postprocess": d => [d[0]].concat(d[1]).filter(notnull)},
    {"name": "line", "symbols": ["_", "command", "eol"], "postprocess": nth(1)},
    {"name": "line", "symbols": ["_", "section", "eol"], "postprocess": d => ({node: 'section', name: d[1]})},
    {"name": "line", "symbols": ["_", "label", "eol"], "postprocess": d => ({node: 'label', name: d[1]})},
    {"name": "line", "symbols": ["eol"], "postprocess": empty},
    {"name": "command$string$1", "symbols": [{"literal":"r"}, {"literal":"a"}, {"literal":"m"}, {"literal":"p"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command", "symbols": ["command$string$1", "__", "float", "sep", "integer"], "postprocess": ([,,time,,amount]) => ({node: 'ramp', time: time, steps: amount})},
    {"name": "command$string$2", "symbols": [{"literal":"w"}, {"literal":"a"}, {"literal":"i"}, {"literal":"t"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command", "symbols": ["command$string$2", "__", "float"], "postprocess": ([,,time]) => ({node: 'ramp', time: time, steps: 0})},
    {"name": "command$string$3", "symbols": [{"literal":"b"}, {"literal":"r"}, {"literal":"a"}, {"literal":"n"}, {"literal":"c"}, {"literal":"h"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command", "symbols": ["command$string$3", "__", "integer", "sep", "labelname"], "postprocess": d => ({node: 'branch', count: d[2], target: d[4]})},
    {"name": "command$string$4", "symbols": [{"literal":"s"}, {"literal":"e"}, {"literal":"t"}, {"literal":"_"}, {"literal":"p"}, {"literal":"w"}, {"literal":"m"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command", "symbols": ["command$string$4", "__", "integer"], "postprocess": ([,,value]) => ({node: 'pwm', value: value})},
    {"name": "command$string$5", "symbols": [{"literal":"s"}, {"literal":"t"}, {"literal":"a"}, {"literal":"r"}, {"literal":"t"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command", "symbols": ["command$string$5"], "postprocess": d => ({node: 'start'})},
    {"name": "command$string$6", "symbols": [{"literal":"t"}, {"literal":"r"}, {"literal":"i"}, {"literal":"g"}, {"literal":"g"}, {"literal":"e"}, {"literal":"r"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command", "symbols": ["command$string$6", "__", "waittarget"], "postprocess": d=> ({node: 'trigger', send: 0, wait: d[2]})},
    {"name": "command$string$7", "symbols": [{"literal":"t"}, {"literal":"r"}, {"literal":"i"}, {"literal":"g"}, {"literal":"g"}, {"literal":"e"}, {"literal":"r"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command$ebnf$1$subexpression$1", "symbols": ["sep", "waittarget"], "postprocess": nth(1)},
    {"name": "command$ebnf$1", "symbols": ["command$ebnf$1$subexpression$1"], "postprocess": id},
    {"name": "command$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "command", "symbols": ["command$string$7", "__", "sendtarget", "command$ebnf$1"], "postprocess": d => ({node: 'trigger', send: d[2] ? d[2] : 0, wait: d[3] ? d[3] : 0})},
    {"name": "command$string$8", "symbols": [{"literal":"e"}, {"literal":"n"}, {"literal":"d"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "command$ebnf$2$subexpression$1$ebnf$1$subexpression$1", "symbols": ["sep", "interruptsignal"], "postprocess": nth(1)},
    {"name": "command$ebnf$2$subexpression$1$ebnf$1", "symbols": ["command$ebnf$2$subexpression$1$ebnf$1$subexpression$1"], "postprocess": id},
    {"name": "command$ebnf$2$subexpression$1$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "command$ebnf$2$subexpression$1", "symbols": ["sep", "interruptsignal", "command$ebnf$2$subexpression$1$ebnf$1"], "postprocess": d => {const s = [d[1], d[2]]; return {reset: s.includes('r'), int: s.includes('i')}}},
    {"name": "command$ebnf$2", "symbols": ["command$ebnf$2$subexpression$1"], "postprocess": id},
    {"name": "command$ebnf$2", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "command", "symbols": ["command$string$8", "command$ebnf$2"], "postprocess": d => (Object.assign({node: 'end', reset: false, int: false}, d[1]))},
    {"name": "sep", "symbols": ["_", {"literal":","}, "_"], "postprocess": empty},
    {"name": "eol$ebnf$1", "symbols": ["comment"], "postprocess": id},
    {"name": "eol$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "eol", "symbols": ["_", "eol$ebnf$1"], "postprocess": empty},
    {"name": "comment$ebnf$1", "symbols": []},
    {"name": "comment$ebnf$1", "symbols": ["comment$ebnf$1", /[^\r\n]/], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "comment", "symbols": [{"literal":"#"}, "comment$ebnf$1"], "postprocess": empty},
    {"name": "label", "symbols": ["labelname", {"literal":":"}], "postprocess": id},
    {"name": "labelname$ebnf$1", "symbols": [/[._a-zA-Z0-9]/]},
    {"name": "labelname$ebnf$1", "symbols": ["labelname$ebnf$1", /[._a-zA-Z0-9]/], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "labelname", "symbols": ["labelname$ebnf$1"], "postprocess": d => d[0].join('')},
    {"name": "interruptsignal", "symbols": [/[rR]/], "postprocess": v("r")},
    {"name": "interruptsignal", "symbols": [/[iI]/], "postprocess": v("i")},
    {"name": "sendtarget", "symbols": [/[sS]/, "target"], "postprocess": nth(1)},
    {"name": "waittarget", "symbols": [/[wW]/, "target"], "postprocess": nth(1)},
    {"name": "target", "symbols": [{"literal":"1"}], "postprocess": v(0b001)},
    {"name": "target", "symbols": [{"literal":"2"}], "postprocess": v(0b010)},
    {"name": "target", "symbols": [{"literal":"3"}], "postprocess": v(0b100)},
    {"name": "target$string$1", "symbols": [{"literal":"1"}, {"literal":"2"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "target", "symbols": ["target$string$1"], "postprocess": v(0b011)},
    {"name": "target$string$2", "symbols": [{"literal":"1"}, {"literal":"3"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "target", "symbols": ["target$string$2"], "postprocess": v(0b101)},
    {"name": "target$string$3", "symbols": [{"literal":"2"}, {"literal":"3"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "target", "symbols": ["target$string$3"], "postprocess": v(0b110)},
    {"name": "target$string$4", "symbols": [{"literal":"1"}, {"literal":"2"}, {"literal":"3"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "target", "symbols": ["target$string$4"], "postprocess": v(0b111)},
    {"name": "section$subexpression$1$string$1", "symbols": [{"literal":"."}, {"literal":"E"}, {"literal":"N"}, {"literal":"G"}, {"literal":"I"}, {"literal":"N"}, {"literal":"E"}, {"literal":"1"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "section$subexpression$1", "symbols": ["section$subexpression$1$string$1"]},
    {"name": "section$subexpression$1$string$2", "symbols": [{"literal":"."}, {"literal":"e"}, {"literal":"n"}, {"literal":"g"}, {"literal":"i"}, {"literal":"n"}, {"literal":"e"}, {"literal":"1"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "section$subexpression$1", "symbols": ["section$subexpression$1$string$2"]},
    {"name": "section", "symbols": ["section$subexpression$1"], "postprocess": v("engine1")},
    {"name": "section$subexpression$2$string$1", "symbols": [{"literal":"."}, {"literal":"E"}, {"literal":"N"}, {"literal":"G"}, {"literal":"I"}, {"literal":"N"}, {"literal":"E"}, {"literal":"2"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "section$subexpression$2", "symbols": ["section$subexpression$2$string$1"]},
    {"name": "section$subexpression$2$string$2", "symbols": [{"literal":"."}, {"literal":"e"}, {"literal":"n"}, {"literal":"g"}, {"literal":"i"}, {"literal":"n"}, {"literal":"e"}, {"literal":"2"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "section$subexpression$2", "symbols": ["section$subexpression$2$string$2"]},
    {"name": "section", "symbols": ["section$subexpression$2"], "postprocess": v("engine2")},
    {"name": "section$subexpression$3$string$1", "symbols": [{"literal":"."}, {"literal":"E"}, {"literal":"N"}, {"literal":"G"}, {"literal":"I"}, {"literal":"N"}, {"literal":"E"}, {"literal":"3"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "section$subexpression$3", "symbols": ["section$subexpression$3$string$1"]},
    {"name": "section$subexpression$3$string$2", "symbols": [{"literal":"."}, {"literal":"e"}, {"literal":"n"}, {"literal":"g"}, {"literal":"i"}, {"literal":"n"}, {"literal":"e"}, {"literal":"3"}], "postprocess": function joiner(d) {return d.join('');}},
    {"name": "section$subexpression$3", "symbols": ["section$subexpression$3$string$2"]},
    {"name": "section", "symbols": ["section$subexpression$3"], "postprocess": v("engine3")},
    {"name": "integer$ebnf$1", "symbols": [{"literal":"-"}], "postprocess": id},
    {"name": "integer$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "integer$ebnf$2", "symbols": [/[0-9]/]},
    {"name": "integer$ebnf$2", "symbols": ["integer$ebnf$2", /[0-9]/], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "integer", "symbols": ["integer$ebnf$1", "integer$ebnf$2"], "postprocess": d => parseInt(d[1].join("") * (d[0] ? -1 : 1))},
    {"name": "float$ebnf$1", "symbols": ["mantissa"], "postprocess": id},
    {"name": "float$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "float", "symbols": ["integer", "float$ebnf$1"], "postprocess": d => Math.sign(d[0]) * (Math.abs(d[0]) + d[1])},
    {"name": "mantissa$ebnf$1", "symbols": []},
    {"name": "mantissa$ebnf$1", "symbols": ["mantissa$ebnf$1", /[0-9]/], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "mantissa", "symbols": [{"literal":"."}, "mantissa$ebnf$1"], "postprocess": d => parseFloat("0"+d.join(''))},
    {"name": "__$ebnf$1", "symbols": ["ws"]},
    {"name": "__$ebnf$1", "symbols": ["__$ebnf$1", "ws"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "__", "symbols": ["__$ebnf$1"], "postprocess": empty},
    {"name": "_$ebnf$1", "symbols": []},
    {"name": "_$ebnf$1", "symbols": ["_$ebnf$1", "ws"], "postprocess": function arrpush(d) {return d[0].concat([d[1]]);}},
    {"name": "_", "symbols": ["_$ebnf$1"], "postprocess": empty},
    {"name": "ws", "symbols": [/[ \t]/], "postprocess": empty},
    {"name": "nl$ebnf$1", "symbols": [{"literal":"\r"}], "postprocess": id},
    {"name": "nl$ebnf$1", "symbols": [], "postprocess": function(d) {return null;}},
    {"name": "nl", "symbols": ["nl$ebnf$1", {"literal":"\n"}], "postprocess": empty}
]
  , ParserStart: "program"
}
if (typeof module !== 'undefined'&& typeof module.exports !== 'undefined') {
   module.exports = grammar;
} else {
   window.grammar = grammar;
}
})();


src/lib/index.ts
import * as nearley from "nearley";
import { Map } from 'core-js';

import * as grammar from './grammar';
import {buf2hex} from './util';

type SectionNames = "engine1" | "engine2" | "engine3";
interface ILabelNode {
  readonly node: "label";
  readonly name: string;
}
interface ISectionNode {
  readonly node: "section";
  readonly name: SectionNames;
}
interface IRampNode {
  readonly node: "ramp";
  readonly time: number;
  readonly steps: number;
}
interface IBranchNode {
  readonly node: "branch";
  readonly count: number;
  readonly target: string;
}
interface IStartNode {
  readonly node: "start";
}
interface IPwmNode {
  readonly node: "pwm";
  readonly value: number;

}
interface IEndNode {
  readonly node: "end";
  readonly reset: boolean;
  readonly int: boolean;
}
interface ITriggerNode {
  readonly node: "trigger";
  readonly send: number;
  readonly wait: number;
}

type INode =
  ILabelNode
  | ISectionNode
  | IPwmNode
  | IRampNode
  | IBranchNode
  | IStartNode
  | IEndNode
  | ITriggerNode;

function throwBadNode(n : never): never;
function throwBadNode(n : INode) {
  throw new Error(`Unknown node type ${n.node}`);
}

class Section {
  public readonly name : string;
  private readonly labels: Map<string, number> = new Map();

  public readonly bitstream : Uint16Array = new Uint16Array(16);
  public readonly dataview: DataView = new DataView(this.bitstream.buffer);
  public readonly number : number;
  private pc : number = 0;

  constructor(name : string) {
    this.name = name;
    this.number = parseInt(name[name.length-1]);
  }

  public addInstruction(ins : number) {
    if(this.pc > 15 && ins != 0) {
      throw new Error(`${this.name} Capacity Exceeded`)
    }
    this.dataview.setInt16(this.pc++ * 2, ins);
  }

  public addLabel(name:string) {
    if(this.labels.has(name)) {
      throw new Error(`Duplicate label "${name}" declared`);
    } 
    this.labels.set(name, this.pc);
  }

  public getLabel(name: string):number {
    let pos = this.labels.get(name);
    if(pos == undefined) throw new Error(`Label "${name}" does not exist`);
    return pos;
  }
}

class LP5562Program {
  public sections: Map<SectionNames, Section> = new Map();
  public clock : number = 32.768;
  private _curSection : Section | null = null;
  private _globalLabelList : Set<string> = new Set();

  private get curSection() : Section {
    if(this._curSection === null) {
      this._curSection = new Section('engine1');
      this.sections.set('engine1', this._curSection);
    }
    return this._curSection;
  }
  private set curSection(section : Section) {
    this._curSection = section;
  }

  public get bitstream() : Uint16Array {
    const bs = new Uint16Array(3*16);
    const e1 = this.sections.get('engine1');
    const e2 = this.sections.get('engine2');
    const e3 = this.sections.get('engine3');

    if(e1) bs.set(e1.bitstream, 0);
    if(e2) bs.set(e2.bitstream, 16);
    if(e3) bs.set(e3.bitstream, 16*2);

    return bs;
  }

  private switchToSection(section : SectionNames) {
    if(this.sections.has(section)) {
      throw new Error(`${section} already declared`);
    } else {
      this._curSection = new Section(section);
      this.sections.set(section, this._curSection);
    }
  }

  public assemble(data: string) : void {
    const p = new nearley.Parser(grammar, grammar.ParserStart);
    const d = p.feed(data);

    if(d.results.length == 0) {
      throw new Error(`Invalid program. (Check that the first instruction is valid?)`);
    }

    const instructions: Array<INode> = d.results[0];
    for (let i of instructions) {
      switch (i.node) {
        case "section":
          this.switchToSection(i.name);
          break;
        case "label":
          const name = i.name;
          if(this._globalLabelList.has(name)) {
            throw new Error(`Duplicate label "${name}" declared`);
          } else {
            this._globalLabelList.add(name);
            this.curSection.addLabel(name);
          }
          break;
        case "trigger":
          this.curSection.addInstruction(
            (0b111 << 13)
            | (i.wait << 7)
            | (i.send << 1)
            );
          break;
        case "end":
          this.curSection.addInstruction(
            (0b110 << 13)
            | (~~i.int << 12)
            | (~~i.reset << 11)
            );
          break;
        case "start":
          this.curSection.addInstruction(0);
          break;
        case "branch":
          if(i.count < 0 || i.count >= 64) {
            throw new Error(`Unsupported loop count: ${i.count}`);
          }

          this.curSection.addInstruction(
            (0b101 << 13)
            | i.count << 7
            | this.curSection.getLabel(i.target)

            // the following is not strictly necessary but is required to match the TI compiler's
            // output since they use absolute position in the DO NOT CARE bits.
            | ((this.curSection.number - 1) & 0b111) << 4
            );
          break;
        case "ramp":
          let steps = Math.abs(i.steps);
          let decrease : boolean = i.steps < 0;
          let increment : number;

          if(steps == 1 || steps > 128) {
            throw new Error(`Invalid number of ramp steps ${steps}`);
          } else if (steps == 0) {
            increment = 0;
          } else {
            increment  = steps - 1;
          }

          let calcCycles = (time:number, clock:number, prescaleFlag:number) => {
              let trueCycles = clock * time;
              let cycles = Math.round(trueCycles);
              let length = cycles / clock;
              let difference = Math.abs(time - length);
              return { cycles, difference, prescaleFlag};
          }

          let guesses = [
            calcCycles(i.time/(increment+1), this.clock / 16, 0),
            calcCycles(i.time/(increment+1), this.clock / 512, 1),
          ];

          guesses = guesses.filter(g => g.cycles < 64 && g.cycles > 0);
          if(guesses.length == 0) {
            throw new Error(`Length ${i.time}ms outside of obtainable range (${1/(this.clock/16)}ms - ${63/(this.clock/512)}ms )`);
          }
          let res = guesses.reduce((prev, curr) => prev.difference < curr.difference ? prev : curr);

          this.curSection.addInstruction(
            (res.prescaleFlag << 14)
            | res.cycles << 8
            | ~~decrease << 7
            | increment
            );
          break;
        case "pwm":
          if(i.value < 0 || i.value > 255) {
            throw new Error(`PWM value ${i.value} out of range`);
          }
          this.curSection.addInstruction(
            (0b010 << 13)
            | i.value
          )
          break;
        default:
          throwBadNode(i);
      }
    }
  }
}


/**
 * Takes in a assembly and returns the full 16 length instructino stream
 * for all 3 engines as a Uint16Array
 */
export function assemble(assembly:string) : Uint16Array  {
  const p = new LP5562Program();
  p.assemble(assembly);
  return p.bitstream;
}
